graph/preprocessor.go

```diff
@@ -0,0 +1,320 @@
+				} else {
+					out.WriteString(resolvedCommand)
+					changed = true
+					if char != alias.directive {
+						ongoingCmd = false
+						out.WriteRune(char)
+					}
+					command.Reset()
+				}
+			} else {
+				command.WriteRune(char)
+			}
+		} else if char == alias.directive {
+			ongoingCmd = true
+		} else {
+			out.WriteRune(char)
+		}
+	}
+
+	return out.String(), changed, nil
+}
+
+// preprocessBytes handles byte encoded data that can be parsed through pre processing
+func preprocessBytes(data []byte) ([]byte, Alias, bool, error) {
+	type Wrapper struct {
+		Alias Alias `yaml:"alias,omitempty"`
+	}
+
+	wrap := &Wrapper{}
+	aliasData, remainingData := basicAliasSeparation(data)
+
+	if errUnmarshal := yaml.Unmarshal(aliasData, wrap); errUnmarshal != nil {
+		return data, Alias{}, false, errUnmarshal
+	}
+
+	alias := &wrap.Alias
+
+	if alias.AliasMap == nil {
+		// Alias Src defined. Guarantees alias map can be populated
+		alias.AliasMap = make(map[string]string)
+	}
+	// Search and Replace
+	str := string(remainingData)
+	parsedStr, changed, err := preprocessString(alias, str)
+
+	return []byte(parsedStr), *alias, changed, err
+}
+
+// processSteps Will resolve image names in steps that are aliased without using directive.
+// Invoked after resolving all directive using aliases
+func processSteps(alias *Alias, task *Task) {
+	for i, step := range task.Steps {
+		parts := strings.Split(strings.TrimSpace(step.Cmd), " ")
+		if val, ok := alias.AliasMap[parts[0]]; ok {
+			// Image name should always go first
+			parts[0] = val
+			task.Steps[i].Cmd = strings.Join(parts, " ")
+		}
+	}
+}
+
+// Provides simple separation of the top level items in a yaml file definition, however, need
+// to update to be fully compliant (include JSON top level for example). Alternative construction
+// of small compiler for this purpose is also under consideration.
+func basicAliasSeparation(data []byte) ([]byte, []byte) {
+	reader := bytes.NewReader(data)
+	scanner := bufio.NewScanner(reader)
+	scanner.Split(bufio.ScanLines)
+
+	var aliasBuffer bytes.Buffer
+	var buffer bytes.Buffer
+
+	inside := false
+	aliasFieldName := regexp.MustCompile(`\Aalias\s*:.*\z`)
+	genericTopLevelRe := regexp.MustCompile(`\A[^\s:]+[^:]*:.*\z`)
+	commentRe := regexp.MustCompile(`\A\s*#.*`)
+	for scanner.Scan() {
+		text := scanner.Text()
+		if matched := commentRe.MatchString(text); matched {
+			continue
+		}
+
+		if matched := aliasFieldName.MatchString(text); matched && !inside {
+			inside = true
+		} else if matched := genericTopLevelRe.MatchString(text); matched && inside {
+			inside = false
+		}
+
+		if inside {
+			aliasBuffer.WriteString(text + "\n")
+		} else {
+			buffer.WriteString(text + "\n")
+		}
+	}
+
+	return aliasBuffer.Bytes(), buffer.Bytes()
+}
+
+// FindVersion Determines the current version of an Alias task file
+func FindVersion(data []byte) string {
+	reader := bytes.NewReader(data)
+	scanner := bufio.NewScanner(reader)
+	scanner.Split(bufio.ScanLines)
```

Comment:
no need. its the default:
https://golang.org/pkg/bufio/#Scanner.Split

remove it

Reply:
May be the default but removing bufio.ScanLines results in build not working. I am gonna keep it then.

Reply:
really? thats odd. We use the similar logic to parse Dockerfiles to build them.

https://github.com/Azure/acr-builder/blob/master/scan/dependencies.go#L164

We also parse that line by line , ignore whitespaces, comments, etc. 

But anyway. We can look at it later. You can keep it for now